File: crates/kuben-api/src/sso.rs
//! The OpenID Connect client for single sign-on (M4.3): discovery, the
//! authorization URL with PKCE, the code exchange and ID token verification.
//!
//! The provider's discovery document must name the configured issuer, and
//! its keys and endpoints are used only from there. ID tokens are accepted
//! only when RS256-signed by one of those keys, for this client and this
//! sign-in's nonce.

use std::{fmt, sync::Arc, time::Duration};

use base64::{Engine as _, engine::general_purpose::URL_SAFE_NO_PAD};
use bytes::Bytes;
use http::{Method, Request, StatusCode, header};
use http_body_util::{BodyExt, Full, Limited};
use kuben_core::{
    config::SsoCfg,
    sso::{IdClaims, SsoDenied, SsoPerson, SsoPolicy},
};
use secrecy::{ExposeSecret, SecretString};
use serde::Deserialize;
use sha2::{Digest as _, Sha256};
use tokio::sync::OnceCell;

use crate::{
    oidc::{HttpGet, HttpsGet, KeyCache, OidcError},
    transport::chain,
};

const TIMEOUT: Duration = Duration::from_secs(15);
const MAX_DOCUMENT: usize = 256 << 10;

/// Why a sign-in failed. People see one answer; the reason is logged.
#[derive(Debug, thiserror::Error)]
pub enum SsoError {
    #[error("the identity provider is unavailable: {0}")]
    Unavailable(String),
    #[error("the identity provider refused the code: {0}")]
    CodeRefused(String),
    #[error("the ID token was not accepted: {0}")]
    Token(#[from] OidcError),
    #[error("{0}")]
    Denied(#[from] SsoDenied),
    #[error("the SSO configuration is incomplete: {0}")]
    Config(String),
}

/// What sign-in needs from the provider, beyond reading documents.
#[async_trait::async_trait]
pub trait IdentityProvider: HttpGet {
    /// `POST` a form to `url` with HTTP Basic client authentication.
    async fn post_form(&self, url: &str, form: &str, basic: &str) -> Result<(StatusCode, Bytes), String>;
}

#[async_trait::async_trait]
impl IdentityProvider for HttpsGet {
    async fn post_form(&self, url: &str, form: &str, basic: &str) -> Result<(StatusCode, Bytes), String> {
        let request = Request::builder()
            .method(Method::POST)
            .uri(url)
            .header(header::ACCEPT, "application/json")
            .header(header::CONTENT_TYPE, "application/x-www-form-urlencoded")
            .header(header::AUTHORIZATION, format!("Basic {basic}"))
            .header(header::USER_AGENT, concat!("kuben/", env!("CARGO_PKG_VERSION")))
            .body(Full::new(Bytes::from(form.to_owned())))
            .map_err(|e| e.to_string())?;
        let response = self.transport().send(request).await?;
        let (parts, body) = response.into_parts();
        let body = tokio::time::timeout(TIMEOUT, Limited::new(body, MAX_DOCUMENT).collect())
            .await
            .map_err(|_| "timed out".to_owned())?
            .map_err(|e| chain(&*e))?
            .to_bytes();
        Ok((parts.status, body))
    }
}

/// The parts of the provider's discovery document Kuben uses.
#[derive(Clone, Debug, Deserialize, PartialEq, Eq)]
pub struct Discovery {
    pub issuer: String,
    pub authorization_endpoint: String,
    pub token_endpoint: String,
    pub jwks_uri: String,
}

struct Provider {
    discovery: Discovery,
    keys: KeyCache,
}

/// Percent-encode a query or form value.
fn encode(value: &str) -> String {
    value
        .bytes()
        .map(|b| {
            if b.is_ascii_alphanumeric() || matches!(b, b'-' | b'.' | b'_' | b'~') {
                char::from(b).to_string()
            } else {
                format!("%{b:02X}")
            }
        })
        .collect()
}

fn https(url: &str) -> bool {
    url.starts_with("https://")
}

/// The PKCE S256 challenge of `verifier`.
#[must_use]
pub fn challenge(verifier: &str) -> String {
    URL_SAFE_NO_PAD.encode(Sha256::digest(verifier.as_bytes()))
}

/// A URL-safe random value of 32 bytes (43 characters).
#[must_use]
pub fn random_value() -> String {
    let bytes: [u8; 32] = rand::random();
    URL_SAFE_NO_PAD.encode(bytes)
}

#[derive(Deserialize)]
struct TokenResponse {
    id_token: Option<String>,
}

/// The single sign-on client of this installation.
pub struct SsoClient {
    issuer: String,
    client_id: String,
    client_secret: SecretString,
    redirect_uri: String,
    scopes: Vec<String>,
    group_claim: String,
    policy: SsoPolicy,
    display_name: String,
    org_slug: String,
    http: Arc<dyn IdentityProvider>,
    provider: OnceCell<Provider>,
    /// Keys given up front (tests).
    fixed_keys: Option<crate::oidc::Jwks>,
    clock: fn() -> i64,
}

impl fmt::Debug for SsoClient {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("SsoClient")
            .field("issuer", &self.issuer)
            .field("client_id", &self.client_id)
            .finish_non_exhaustive()
    }
}

fn system_clock() -> i64 {
    kuben_core::time::now_ms() / 1000
}

impl SsoClient {
    /// The client `cfg` describes, answering at `public_url`, joining people
    /// to `default_org` unless the configuration names another.
    ///
    /// # Errors
    ///
    /// When a required setting is missing or invalid.
    pub fn from_config(cfg: &SsoCfg, public_url: Option<&str>, default_org: &str) -> Result<Self, SsoError> {
        let missing = |what: &str| SsoError::Config(format!("sso.{what} is required"));
        let issuer = cfg.issuer.clone().ok_or_else(|| missing("issuer"))?;
        let client_id = cfg.client_id.clone().ok_or_else(|| missing("client_id"))?;
        let secret = match (&cfg.client_secret_file, &cfg.client_secret) {
            (Some(path), _) => std::fs::read_to_string(path)
                .map_err(|e| SsoError::Config(format!("cannot read sso.client_secret_file {path}: {e}")))?
                .trim()
                .to_owned(),
            (None, Some(secret)) => secret.clone(),
            (None, None) => return Err(missing("client_secret_file")),
        };
        let public_url = public_url
            .filter(|u| https(u))
            .ok_or_else(|| SsoError::Config("server.public_url must be an https URL for SSO".into()))?;
        let policy = cfg.policy().map_err(|e| SsoError::Config(e.to_string()))?;
        if policy.groups.is_empty() && policy.default_role.is_none() {
            return Err(SsoError::Config(
                "map at least one group to a role (sso.groups) or set sso.default_role".into(),
            ));
        }
        Ok(Self {
            issuer: issuer.trim_end_matches('/').to_owned(),
            client_id,
            client_secret: SecretString::from(secret),
            redirect_uri: format!("{}/api/v1/auth/sso/callback", public_url.trim_end_matches('/')),
            scopes: cfg.scopes.clone(),
            group_claim: cfg.group_claim.clone(),
            policy,
            display_name: cfg.display_name.clone(),
            org_slug: cfg.org.clone().unwrap_or_else(|| default_org.to_owned()),
            http: Arc::new(HttpsGet::default()),
            provider: OnceCell::new(),
            fixed_keys: None,
            clock: system_clock,
        })
    }

    /// The same client talking to `http` with fixed keys and `clock` (tests).
    #[must_use]
    pub fn with_provider(
        mut self,
        http: Arc<dyn IdentityProvider>,
        keys: crate::oidc::Jwks,
        clock: fn() -> i64,
    ) -> Self {
        self.http = http;
        self.fixed_keys = Some(keys);
        self.clock = clock;
        self
    }

    #[must_use]
    pub fn display_name(&self) -> &str {
        &self.display_name
    }

    #[must_use]
    pub fn org_slug(&self) -> &str {
        &self.org_slug
    }

    #[must_use]
    pub fn issuer(&self) -> &str {
        &self.issuer
    }

    async fn provider(&self) -> Result<&Provider, SsoError> {
        self.provider
            .get_or_try_init(|| async {
                let url = format!("{}/.well-known/openid-configuration", self.issuer);
                let body = self
                    .http
                    .get(&url, MAX_DOCUMENT)
                    .await
                    .map_err(SsoError::Unavailable)?;
                let discovery: Discovery =
                    serde_json::from_slice(&body).map_err(|e| SsoError::Unavailable(e.to_string()))?;
                if discovery.issuer.trim_end_matches('/') != self.issuer {
                    return Err(SsoError::Config(format!(
                        "the provider calls itself {}, not {}",
                        discovery.issuer, self.issuer
                    )));
                }
                let endpoints = [
                    &discovery.authorization_endpoint,
                    &discovery.token_endpoint,
                    &discovery.jwks_uri,
                ];
                if !endpoints.iter().all(|u| https(u)) {
                    return Err(SsoError::Config("the provider's endpoints must be https".into()));
                }
                let keys = match &self.fixed_keys {
                    Some(jwks) => KeyCache::fixed(jwks.clone()),
                    None => KeyCache::new(discovery.jwks_uri.clone(), self.http.clone() as Arc<dyn HttpGet>),
                };
                Ok(Provider { discovery, keys })
            })
            .await
    }

    /// Where to send the browser to sign in.
    pub async fn authorize_url(&self, state: &str, nonce: &str, verifier: &str) -> Result<String, SsoError> {
        let p = self.provider().await?;
        let separator = if p.discovery.authorization_endpoint.contains('?') {
            '&'
        } else {
            '?'
        };
        Ok(format!(
            "{}{separator}response_type=code&client_id={}&redirect_uri={}&scope={}&state={}&nonce={}\
             &code_challenge={}&code_challenge_method=S256",
            p.discovery.authorization_endpoint,
            encode(&self.client_id),
            encode(&self.redirect_uri),
            encode(&self.scopes.join(" ")),
            encode(state),
            encode(nonce),
            challenge(verifier),
        ))
    }

    /// The person behind `code`, if the provider and the policy admit them.
    pub async fn sign_in(&self, code: &str, verifier: &str, nonce: &str) -> Result<SsoPerson, SsoError> {
        let p = self.provider().await?;
        let form = format!(
            "grant_type=authorization_code&code={}&redirect_uri={}&code_verifier={}&client_id={}",
            encode(code),
            encode(&self.redirect_uri),
            encode(verifier),
            encode(&self.client_id),
        );
        let basic = base64::engine::general_purpose::STANDARD.encode(format!(
            "{}:{}",
            encode(&self.client_id),
            encode(self.client_secret.expose_secret())
        ));
        let (status, body) = self
            .http
            .post_form(&p.discovery.token_endpoint, &form, &basic)
            .await
            .map_err(SsoError::Unavailable)?;
        if !status.is_success() {
            return Err(SsoError::CodeRefused(format!("HTTP {status}")));
        }
        let response: TokenResponse =
            serde_json::from_slice(&body).map_err(|e| SsoError::CodeRefused(e.to_string()))?;
        let id_token = response
            .id_token
            .ok_or_else(|| SsoError::CodeRefused("no id_token".into()))?;
        let payload = p.keys.verify(&id_token).await?;
        let claims: IdClaims = serde_json::from_slice(&payload).map_err(|_| OidcError::Malformed)?;
        claims.check(&self.issuer, &self.client_id, nonce, (self.clock)())?;
        let groups = claims.groups(&self.group_claim);
        Ok(self.policy.admit(&claims, &groups)?)
    }
}

#[cfg(test)]
pub(crate) mod tests {
    use std::{collections::BTreeMap, sync::Mutex};

    use kuben_core::perm::Role;

    use super::*;

    pub(crate) const FIXTURE: &str = include_str!("testdata/sso-oidc.json");
    pub(crate) const NOW: i64 = 1_800_000_100;

    /// A provider answering from memory; the token endpoint hands out the
    /// fixture token named by the code.
    #[derive(Default)]
    pub(crate) struct FakeIdp {
        pub(crate) forms: Mutex<Vec<String>>,
        pub(crate) issuer: Option<String>,
    }

    #[async_trait::async_trait]
    impl HttpGet for FakeIdp {
        async fn get(&self, url: &str, _limit: usize) -> Result<Bytes, String> {
            assert_eq!(url, "https://idp.example.com/.well-known/openid-configuration");
            let issuer = self
                .issuer
                .clone()
                .unwrap_or_else(|| "https://idp.example.com".into());
            Ok(Bytes::from(
                serde_json::json!({
                    "issuer": issuer,
                    "authorization_endpoint": "https://idp.example.com/authorize",
                    "token_endpoint": "https://idp.example.com/token",
                    "jwks_uri": "https://idp.example.com/jwks",
                })
                .to_string(),
            ))
        }
    }

    #[async_trait::async_trait]
    impl IdentityProvider for FakeIdp {
        async fn post_form(&self, url: &str, form: &str, basic: &str) -> Result<(StatusCode, Bytes), String> {
            assert_eq!(url, "https://idp.example.com/token");
            assert!(!basic.is_empty());
            self.forms.lock().expect("lock").push(form.to_owned());
            let code = form
                .split('&')
                .find_map(|kv| kv.strip_prefix("code="))
                .unwrap_or_default();
            let tokens = fixture()["tokens"].clone();
            match tokens.get(code).and_then(|t| t.as_str()) {
                Some(token) => Ok((
                    StatusCode::OK,
                    Bytes::from(serde_json::json!({ "id_token": token, "token_type": "Bearer" }).to_string()),
                )),
                None => Ok((
                    StatusCode::BAD_REQUEST,
                    Bytes::from_static(b"{\"error\":\"invalid_grant\"}"),
                )),
            }
        }
    }

    pub(crate) fn fixture() -> serde_json::Value {
        serde_json::from_str(FIXTURE).expect("fixture")
    }

    pub(crate) fn config() -> SsoCfg {
        SsoCfg {
            enabled: true,
            issuer: Some("https://idp.example.com/".into()),
            client_id: Some("kuben-console".into()),
            client_secret: Some("s3cret".into()),
            groups: BTreeMap::from([
                ("platform-admins".into(), "admin".into()),
                ("devs".into(), "developer".into()),
            ]),
            allowed_domains: vec!["example.com".into()],
            ..SsoCfg::default()
        }
    }

    pub(crate) fn client(idp: Arc<FakeIdp>) -> SsoClient {
        let jwks = serde_json::from_value(fixture()["jwks"].clone()).expect("jwks");
        SsoClient::from_config(&config(), Some("https://kuben.example.com/"), "acme")
            .expect("client")
            .with_provider(idp, jwks, || NOW)
    }

    #[test]
    fn configuration_is_checked_up_front() {
        let ok =
            SsoClient::from_config(&config(), Some("https://kuben.example.com"), "acme").expect("client");
        assert_eq!(ok.org_slug(), "acme");
        assert_eq!(ok.issuer(), "https://idp.example.com");
        assert_eq!(
            ok.redirect_uri,
            "https://kuben.example.com/api/v1/auth/sso/callback"
        );
        assert!(!format!("{ok:?}").contains("s3cret"));
        let cases = [
            (
                SsoCfg {
                    issuer: None,
                    ..config()
                },
                Some("https://kuben.example.com"),
            ),
            (
                SsoCfg {
                    client_secret: None,
                    ..config()
                },
                Some("https://kuben.example.com"),
            ),
            (config(), Some("http://kuben.example.com")),
            (config(), None),
            (
                SsoCfg {
                    groups: BTreeMap::new(),
                    ..config()
                },
                Some("https://kuben.example.com"),
            ),
        ];
        for (cfg, url) in cases {
            assert!(matches!(
                SsoClient::from_config(&cfg, url, "acme"),
                Err(SsoError::Config(_))
            ));
        }
        let defaulted = SsoCfg {
            groups: BTreeMap::new(),
            default_role: Some("viewer".into()),
            ..config()
        };
        assert!(SsoClient::from_config(&defaulted, Some("https://k.example.com"), "acme").is_ok());
    }

    #[tokio::test]
    async fn the_authorization_url_carries_pkce_and_the_nonce() {
        let c = client(Arc::default());
        let url = c
            .authorize_url("st@te", "nonce-1", "verifier-verifier-verifier-verifier-verif")
            .await
            .expect("url");
        assert!(
            url.starts_with("https://idp.example.com/authorize?response_type=code&client_id=kuben-console")
        );
        assert!(
            url.contains("redirect_uri=https%3A%2F%2Fkuben.example.com%2Fapi%2Fv1%2Fauth%2Fsso%2Fcallback")
        );
        assert!(url.contains("scope=openid%20email%20profile"));
        assert!(url.contains("state=st%40te") && url.contains("nonce=nonce-1"));
        assert!(url.contains(&format!(
            "code_challenge={}&code_challenge_method=S256",
            challenge("verifier-verifier-verifier-verifier-verif")
        )));
        assert_eq!(
            challenge("dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk"),
            "E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM"
        );
        assert_eq!(random_value().len(), 43);
        assert_ne!(random_value(), random_value());
    }

    #[tokio::test]
    async fn a_valid_code_signs_the_mapped_person_in() {
        let idp = Arc::new(FakeIdp::default());
        let c = client(idp.clone());
        let carol = c
            .sign_in("valid", "the-verifier", "nonce-1")
            .await
            .expect("carol");
        assert_eq!(
            (carol.email.as_str(), carol.role),
            ("carol@example.com", Role::Admin)
        );
        let dave = c.sign_in("developer", "v", "nonce-1").await.expect("dave");
        assert_eq!(dave.role, Role::Developer);
        let form = idp.forms.lock().expect("lock")[0].clone();
        assert!(
            form.contains("grant_type=authorization_code") && form.contains("code_verifier=the-verifier")
        );
    }

    #[tokio::test]
    async fn anything_else_is_refused() {
        let c = client(Arc::default());
        let denied = |r: Result<SsoPerson, SsoError>| match r {
            Err(SsoError::Denied(d)) => Some(d),
            _ => None,
        };
        assert_eq!(
            denied(c.sign_in("valid", "v", "nonce-2").await),
            Some(SsoDenied::Nonce)
        );
        assert_eq!(
            denied(c.sign_in("unverified", "v", "nonce-1").await),
            Some(SsoDenied::Unverified)
        );
        assert_eq!(
            denied(c.sign_in("outsider", "v", "nonce-1").await),
            Some(SsoDenied::Domain)
        );
        assert_eq!(
            denied(c.sign_in("no_role", "v", "nonce-1").await),
            Some(SsoDenied::NoRole)
        );
        assert_eq!(
            denied(c.sign_in("wrong_aud", "v", "nonce-1").await),
            Some(SsoDenied::Audience)
        );
        assert!(matches!(
            c.sign_in("unknown", "v", "nonce-1").await,
            Err(SsoError::CodeRefused(_))
        ));
    }

    #[tokio::test]
    async fn a_provider_claiming_another_issuer_is_not_used() {
        let idp = Arc::new(FakeIdp {
            issuer: Some("https://evil.example.com".into()),
            ..FakeIdp::default()
        });
        let c = client(idp);
        assert!(matches!(
            c.sign_in("valid", "v", "nonce-1").await,
            Err(SsoError::Config(_))
        ));
    }
}

File: internal/integrations/sso/sso.go
// Package sso is the OpenID Connect client for single sign-on (M4.3):
// discovery, the authorization URL with PKCE, the code exchange and ID
// token verification. It replaces crates/kuben-api/src/sso.rs; tokens are
// verified by package oidc (RS256 over the provider's JWKS), the claims and
// the policy are package core/sso.
//
// The provider's discovery document must name the configured issuer, and
// its keys and endpoints are used only from there. ID tokens are accepted
// only when RS256-signed by one of those keys, for this client and this
// sign-in's nonce.
package sso

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strings"
	"sync"

	"github.com/Teamtem-dev/kuben/internal/core/clock"
	"github.com/Teamtem-dev/kuben/internal/core/config"
	"github.com/Teamtem-dev/kuben/internal/core/opt"
	coresso "github.com/Teamtem-dev/kuben/internal/core/sso"
	"github.com/Teamtem-dev/kuben/internal/integrations/oidc"
	"github.com/Teamtem-dev/kuben/internal/integrations/outbound"
	"github.com/Teamtem-dev/kuben/internal/jsonx"
)

// MaxDocument is the largest answer read from the provider, in bytes.
const MaxDocument = 256 << 10

// Why a sign-in failed. People see one answer; the reason is logged. A
// person or token the policy refuses fails with the [coresso.Denied]
// itself.
type (
	// UnavailableError is a provider that cannot be reached or read.
	UnavailableError struct{ Reason string }
	// CodeRefusedError is a code the token endpoint did not exchange.
	CodeRefusedError struct{ Reason string }
	// TokenError is an ID token that did not verify (an oidc error).
	TokenError struct{ Err error }
	// ConfigError is a configuration that cannot work.
	ConfigError struct{ Reason string }
)

func (e UnavailableError) Error() string {
	return "the identity provider is unavailable: " + e.Reason
}

func (e CodeRefusedError) Error() string {
	return "the identity provider refused the code: " + e.Reason
}

func (e TokenError) Error() string { return "the ID token was not accepted: " + e.Err.Error() }

// Unwrap is the oidc error.
func (e TokenError) Unwrap() error { return e.Err }

func (e ConfigError) Error() string {
	return "the SSO configuration is incomplete: " + e.Reason
}

// IdentityProvider is what sign-in needs from the provider: reading its
// documents, and posting a form to its token endpoint.
type IdentityProvider interface {
	// Get is the body of url if it answers 200, at most limit bytes.
	Get(ctx context.Context, url string, limit int64) ([]byte, error)
	// PostForm posts form to url with HTTP Basic client authentication
	// basic; the answer's status and body, whatever the status.
	PostForm(ctx context.Context, url, form, basic string) (int, []byte, error)
}

// Discovery is the parts of the provider's discovery document Kuben uses.
type Discovery struct {
	Issuer                string
	AuthorizationEndpoint string
	TokenEndpoint         string
	JWKSURI               string
}

// UnmarshalJSON reads a discovery document: the four members must be
// there, the rest is ignored.
func (d *Discovery) UnmarshalJSON(data []byte) error {
	var o jsonx.Object
	if err := json.Unmarshal(data, &o); err != nil {
		return err //nolint:wrapcheck // the JSON error is the answer
	}
	var out Discovery
	for _, err := range []error{
		jsonx.Required(o, "issuer", &out.Issuer),
		jsonx.Required(o, "authorization_endpoint", &out.AuthorizationEndpoint),
		jsonx.Required(o, "token_endpoint", &out.TokenEndpoint),
		jsonx.Required(o, "jwks_uri", &out.JWKSURI),
	} {
		if err != nil {
			return err
		}
	}
	*d = out
	return nil
}

type provider struct {
	discovery Discovery
	keys      *oidc.KeyCache
}

// Encode percent-encodes a query or form value: everything but ASCII
// letters, digits and `-._~`, in upper-case hex.
func Encode(value string) string {
	var b strings.Builder
	for i := range len(value) {
		c := value[i]
		if 'a' <= c && c <= 'z' || 'A' <= c && c <= 'Z' || '0' <= c && c <= '9' || strings.IndexByte("-._~", c) >= 0 {
			b.WriteByte(c)
			continue
		}
		fmt.Fprintf(&b, "%%%02X", c)
	}
	return b.String()
}

func https(url string) bool { return strings.HasPrefix(url, "https://") }

// Challenge is the PKCE S256 challenge of verifier.
func Challenge(verifier string) string {
	sum := sha256.Sum256([]byte(verifier))
	return base64.RawURLEncoding.EncodeToString(sum[:])
}

// RandomValue is a URL-safe random value of 32 bytes (43 characters).
func RandomValue() string {
	var b [32]byte
	_, _ = rand.Read(b[:]) //nolint:errcheck // crypto/rand.Read never fails (Go ≥ 1.24)
	return base64.RawURLEncoding.EncodeToString(b[:])
}

// Client is the single sign-on client of this installation.
type Client struct {
	issuer       string
	clientID     string
	clientSecret config.Secret
	redirectURI  string
	scopes       []string
	groupClaim   string
	policy       coresso.Policy
	displayName  string
	orgSlug      string
	http         IdentityProvider
	// outbound reads the provider's keys; nil with fixed keys.
	outbound *outbound.Client
	// fixedKeys are keys given up front (tests).
	fixedKeys opt.Val[oidc.JWKS]
	clock     clock.Clock

	// mu guards provider; it is held while the provider is discovered,
	// so concurrent sign-ins wait for one discovery (Rust's OnceCell).
	mu       sync.Mutex
	provider opt.Val[provider]
}

// FromConfig is the client cfg describes, answering at publicURL, joining
// people to defaultOrg unless the configuration names another. The error
// is a [ConfigError] when a required setting is missing or invalid.
func FromConfig(cfg config.SsoCfg, publicURL opt.Val[string], defaultOrg string) (*Client, error) {
	missing := func(what string) error { return ConfigError{Reason: "sso." + what + " is required"} }
	issuer, ok := cfg.Issuer.Get()
	if !ok {
		return nil, missing("issuer")
	}
	clientID, ok := cfg.ClientID.Get()
	if !ok {
		return nil, missing("client_id")
	}
	var secret config.Secret
	switch path, fromFile := cfg.ClientSecretFile.Get(); {
	case fromFile:
		text, err := os.ReadFile(path) //nolint:gosec // an operator-configured path
		if err != nil {
			return nil, ConfigError{Reason: fmt.Sprintf("cannot read sso.client_secret_file %s: %v", path, err)}
		}
		secret = config.Secret(strings.TrimSpace(string(text)))
	case cfg.ClientSecret.IsSet():
		secret = cfg.ClientSecret
	default:
		return nil, missing("client_secret_file")
	}
	public, ok := publicURL.Get()
	if !ok || !https(public) {
		return nil, ConfigError{Reason: "server.public_url must be an https URL for SSO"}
	}
	policy, err := cfg.Policy()
	if err != nil {
		return nil, ConfigError{Reason: err.Error()}
	}
	if len(policy.Groups) == 0 && policy.DefaultRole.IsNone() {
		return nil, ConfigError{Reason: "map at least one group to a role (sso.groups) or set sso.default_role"}
	}
	client := outbound.New(false, oidc.Timeout)
	return &Client{
		issuer:       strings.TrimRight(issuer, "/"),
		clientID:     clientID,
		clientSecret: secret,
		redirectURI:  strings.TrimRight(public, "/") + "/api/v1/auth/sso/callback",
		scopes:       cfg.Scopes,
		groupClaim:   cfg.GroupClaim,
		policy:       policy,
		displayName:  cfg.DisplayName,
		orgSlug:      cfg.Org.Or(defaultOrg),
		http:         HTTPS{Client: client},
		outbound:     client,
		clock:        clock.System{},
	}, nil
}

// WithProvider is the same client talking to idp with fixed keys, reading
// the time from clk (tests). Call it before the client is used.
func (c *Client) WithProvider(idp IdentityProvider, keys oidc.JWKS, clk clock.Clock) *Client {
	c.http, c.fixedKeys, c.clock, c.outbound = idp, opt.Some(keys), clk, nil
	return c
}

// Format shows the issuer and the client id only, whatever the verb: the
// client secret never reaches a log (Rust's finish_non_exhaustive Debug).
func (c *Client) Format(f fmt.State, _ rune) {
	_, _ = fmt.Fprintf(f, "sso.Client{issuer: %q, clientID: %q, ..}", c.issuer, c.clientID) //nolint:errcheck // fmt.State cannot report
}

// DisplayName is the sign-in button's label.
func (c *Client) DisplayName() string { return c.displayName }

// OrgSlug is the organization people join.
func (c *Client) OrgSlug() string { return c.orgSlug }

// Issuer is the provider's issuer, without trailing slashes.
func (c *Client) Issuer() string { return c.issuer }

// RedirectURI is where the provider sends the browser back.
func (c *Client) RedirectURI() string { return c.redirectURI }

// discover is the provider, read from its discovery document once.
func (c *Client) discover(ctx context.Context) (provider, error) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if p, ok := c.provider.Get(); ok {
		return p, nil
	}
	body, err := c.http.Get(ctx, c.issuer+"/.well-known/openid-configuration", MaxDocument)
	if err != nil {
		return provider{}, UnavailableError{Reason: err.Error()}
	}
	var d Discovery
	if err := json.Unmarshal(body, &d); err != nil {
		return provider{}, UnavailableError{Reason: err.Error()}
	}
	if strings.TrimRight(d.Issuer, "/") != c.issuer {
		return provider{}, ConfigError{Reason: fmt.Sprintf("the provider calls itself %s, not %s", d.Issuer, c.issuer)}
	}
	if !https(d.AuthorizationEndpoint) || !https(d.TokenEndpoint) || !https(d.JWKSURI) {
		return provider{}, ConfigError{Reason: "the provider's endpoints must be https"}
	}
	keys := oidc.NewKeyCache(d.JWKSURI, c.outbound, c.clock, nil)
	if jwks, ok := c.fixedKeys.Get(); ok {
		keys = oidc.FixedKeys(jwks)
	}
	p := provider{discovery: d, keys: keys}
	c.provider = opt.Some(p)
	return p, nil
}

// AuthorizeURL is where to send the browser to sign in.
func (c *Client) AuthorizeURL(ctx context.Context, state, nonce, verifier string) (string, error) {
	p, err := c.discover(ctx)
	if err != nil {
		return "", err
	}
	separator := "?"
	if strings.Contains(p.discovery.AuthorizationEndpoint, "?") {
		separator = "&"
	}
	return p.discovery.AuthorizationEndpoint + separator +
		"response_type=code&client_id=" + Encode(c.clientID) +
		"&redirect_uri=" + Encode(c.redirectURI) +
		"&scope=" + Encode(strings.Join(c.scopes, " ")) +
		"&state=" + Encode(state) +
		"&nonce=" + Encode(nonce) +
		"&code_challenge=" + Challenge(verifier) + "&code_challenge_method=S256", nil
}

// SignIn is the person behind code, if the provider and the policy admit
// them.
func (c *Client) SignIn(ctx context.Context, code, verifier, nonce string) (coresso.Person, error) {
	p, err := c.discover(ctx)
	if err != nil {
		return coresso.Person{}, err
	}
	idToken, err := c.exchange(ctx, p, code, verifier)
	if err != nil {
		return coresso.Person{}, err
	}
	payload, err := p.keys.Verify(ctx, idToken)
	if err != nil {
		return coresso.Person{}, TokenError{Err: err}
	}
	var claims coresso.IDClaims
	if json.Unmarshal(payload, &claims) != nil {
		return coresso.Person{}, TokenError{Err: oidc.ErrMalformed}
	}
	if err := claims.Check(c.issuer, c.clientID, nonce, c.clock.NowMs()/1000); err != nil {
		return coresso.Person{}, err //nolint:wrapcheck // a coresso.Denied, matched on
	}
	person, err := c.policy.Admit(claims, claims.Groups(c.groupClaim))
	if err != nil {
		return coresso.Person{}, err //nolint:wrapcheck // a coresso.Denied, matched on
	}
	return person, nil
}

// exchange trades code for an ID token at the token endpoint.
func (c *Client) exchange(ctx context.Context, p provider, code, verifier string) (string, error) {
	form := "grant_type=authorization_code&code=" + Encode(code) +
		"&redirect_uri=" + Encode(c.redirectURI) +
		"&code_verifier=" + Encode(verifier) +
		"&client_id=" + Encode(c.clientID)
	basic := base64.StdEncoding.EncodeToString([]byte(Encode(c.clientID) + ":" + Encode(c.clientSecret.Expose())))
	status, body, err := c.http.PostForm(ctx, p.discovery.TokenEndpoint, form, basic)
	if err != nil {
		return "", UnavailableError{Reason: err.Error()}
	}
	if status < 200 || status > 299 {
		return "", CodeRefusedError{Reason: httpStatus(status)}
	}
	var o jsonx.Object
	var idToken opt.Val[string]
	if err := json.Unmarshal(body, &o); err != nil {
		return "", CodeRefusedError{Reason: err.Error()}
	}
	if err := jsonx.Optional(o, "id_token", &idToken); err != nil {
		return "", CodeRefusedError{Reason: err.Error()}
	}
	token, ok := idToken.Get()
	if !ok {
		return "", CodeRefusedError{Reason: "no id_token"}
	}
	return token, nil
}

// httpStatus is a status as http's StatusCode displayed it: `HTTP 400 Bad Request`.
func httpStatus(status int) string {
	return strings.TrimSpace(fmt.Sprintf("HTTP %d %s", status, http.StatusText(status)))
}
